File: 74-search-a-2d-matrix/search-a-2d-matrix.ts
function searchMatrix(matrix: number[][], target: number): boolean {
    if (matrix.length === 0 || matrix[0].length === 0) return false

    const rows: number = matrix.length
    const cols: number = matrix[0].length

    let left: number = 0
    let right: number = rows * cols - 1

    while (left <= right) {
        const mid: number = Math.floor((left + right) / 2)
        const midElement: number = matrix[Math.floor(mid / cols)][mid % cols]

        if (midElement === target) {
            return true
        } else if (midElement < target) {
            left = mid + 1
        } else {
            right = mid - 1
        }
    }

    return false
};

File: 74-search-a-2d-matrix/search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if not matrix or not matrix[0]:
            return False

        rows: int = len(matrix)
        cols: int = len(matrix[0])

        left: int = 0
        right: int = rows * cols - 1

        while left <= right:
            mid: int = (left + right) // 2
            row: int = mid // cols
            col: int = mid % cols
            mid_element: int = matrix[row][col]

            if mid_element == target:
                return True
            elif mid_element < target:
                left = mid + 1
            else:
                right = mid - 1
        return False